src/state.rs

```diff
@@ -0,0 +1,214 @@
+///         errors: vec![
+///             Error::Unexpected('8'.into()),
+///             Error::Expected('9'.into()),
+///             Error::Message("Not a nine".into())
+///         ]
+///     }));
+/// # }
+/// ```
+#[derive(Clone, PartialEq)]
+pub struct State<I, X>
+    where I: Stream,
+          X: Positioner<Position = I::Position>,
+          I::Position: Clone + Ord
+{
+    /// The input stream used when items are requested
+    pub input: I,
+    /// The positioner used to update the current position
+    pub positioner: X,
+}
+
+impl<I, X> fmt::Debug for State<I, X>
+    where I: Stream + fmt::Debug,
+          X: Positioner<Position = I::Position>,
+          I::Position: Clone + Ord + fmt::Debug
+{
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f,
+               "State {{ position: {:?}, input: {:?} }}",
+               self.positioner.position(),
+               self.input)
+    }
+}
+
+impl<I, X> State<I, X>
+    where I: Stream,
+          X: Positioner<Position = I::Position>,
+          I::Position: Clone + Ord
+{
+    /// Creates a new `State<I, X>` from an input stream and a positioner.
+    pub fn new(input: I, positioner: X) -> State<I, X> {
+        State {
+            input: input,
+            positioner: positioner,
+        }
+    }
+}
+
+impl<I, X> StreamOnce for State<I, X>
+    where I: Stream,
+          X: Positioner<Position = I::Position, Item = I::Item>,
+          I::Position: Clone + Ord
+{
+    type Item = I::Item;
+    type Range = I::Range;
+    type Position = I::Position;
+
+    #[inline]
+    fn uncons(&mut self) -> Result<I::Item, Error<I::Item, I::Range>> {
+        match self.input.uncons() {
+            Ok(c) => {
+                self.positioner.update(&c);
+                Ok(c)
+            }
+            Err(err) => Err(err),
+        }
+    }
+
+    #[inline(always)]
+    fn position(&self) -> Self::Position {
+        self.positioner.position()
+    }
+}
+
+/// The `IndexPositioner<Item, Range>` struct maintains the current index into the stream `I`.  The
+/// initial index is index 0.  Each `Item` consumed increments the index by 1; each `range` consumed
+/// increments the position by `range.len()`.
+#[derive(Clone, PartialEq)]
+pub struct IndexPositioner<Item, Range>(usize,
+                                        PhantomData<Item>,
+                                        PhantomData<Range>);
+
+impl<Item, Range> Positioner for IndexPositioner<Item, Range>
+    where Item: PartialEq + Clone,
+          Range: PartialEq + Clone + ::primitives::Range
+{
+    type Item = Item;
+    type Position = usize;
+
+    #[inline(always)]
+    fn position(&self) -> usize {
+        self.0
+    }
+
+    #[inline]
+    fn update(&mut self, _item: &Self::Item) {
+        self.0 += 1
+    }
+}
+
+impl<Item, Range> IndexPositioner<Item, Range> {
+    pub fn new() -> IndexPositioner<Item, Range> {
+        IndexPositioner::<Item, Range>::new_with_position(0)
```

Comment:
No need for `::<Item, Range>::` type inference can figure out the correct types.